### PR TITLE
Eliminate GenericSignature::enumeratePairedRequirements()

### PR DESCRIPTION
**Pull request by DougGregor**
This method's primary use was to form `SubstitutionList`s. Those are gone now, so migrate all of the callers away and eliminate this complicated code.

### CODE REVIEW

**Comment by slavapestov** [comment]
Excellent!

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/AST/SubstitutionMapStorage.h [hunk 1]
```diff
@@ -35,7 +35,10 @@ class SubstitutionMap::Storage final
 
   /// The number of conformance requirements, cached to avoid constantly
   /// recomputing it on conformance-buffer access.
-  const unsigned numConformanceRequirements;
+  const unsigned numConformanceRequirements : 31;
+
+  /// Whether we've populated all replacement types already.
+  bool populatedAllReplacements = false;
```
Comment (slavapestov) [inline lib/AST/SubstitutionMapStorage.h]: Does this need to be `: 1`?

File: lib/AST/SubstitutionMap.cpp [hunk 2]
```diff
@@ -80,8 +80,12 @@ ArrayRef<Type> SubstitutionMap::getReplacementTypes() const {
   if (empty()) return { };
 
   // Make sure we've filled in all of the replacement types.
-  for (auto gp : getGenericSignature()->getGenericParams()) {
-    (void)Type(gp).subst(*this);
+  if (!storage->populatedAllReplacements) {
+    for (auto gp : getGenericSignature()->getGenericParams()) {
+      (void)Type(gp).subst(*this);
```
Comment (slavapestov) [inline lib/AST/SubstitutionMap.cpp]: Slightly more efficient to call `subMap.lookupSubstitution(gp)` I think

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Pull request merged**